File: typescript/tutorials/tour-of-workflows-typescript/src/workflows/signup-with-events.ts
import * as restate from "@restatedev/restate-sdk";
import {
  WorkflowContext,
  WorkflowSharedContext,
} from "@restatedev/restate-sdk";
import { createUser, activateUser, sendWelcomeEmail, User } from "../utils";

export const signupWithEvents = restate.workflow({
  name: "SignupWithEventsWorkflow",
  handlers: {
    run: async (ctx: WorkflowContext, user: User) => {
      const userId = ctx.key;

      const success = await ctx.run("create", () => createUser(userId, user));
      if (!success) {
        await ctx.promise<string>("user-created").reject("Creation failed.");
        return { success };
      }
      await ctx.promise<string>("user-created").resolve("User created.");

      await ctx.run("activate", () => activateUser(userId));
      await ctx.run("welcome", () => sendWelcomeEmail(user));
      return { success };
    },

    waitForUserCreation: async (ctx: WorkflowSharedContext) => {
      return ctx.promise<string>("user-created");
    },
  },
});

File: python/tutorials/tour-of-workflows-python/app/workflows/signup_with_events.py
import restate
from restate import WorkflowContext, WorkflowSharedContext
from app.utils import User, create_user, activate_user, send_welcome_email

signup_with_events = restate.Workflow("SignupWithEventsWorkflow")


@signup_with_events.main()
async def run(ctx: WorkflowContext, user: User) -> bool:
    user_id = ctx.key()

    success = await ctx.run_typed("create", create_user, user_id=user_id, user=user)
    if not success:
        await ctx.promise("user-created").reject("Creation failed.")
        return False

    await ctx.promise("user-created").resolve("User created.")

    await ctx.run_typed("activate", activate_user, user_id=user_id)
    await ctx.run_typed("welcome", send_welcome_email, user=user)
    return True


@signup_with_events.handler("waitForUserCreation")
async def wait_for_user_creation(ctx: WorkflowSharedContext) -> str:
    return await ctx.promise("user-created").value()
